### PR TITLE
feat(agent): agent controller upgrade

### PR DESCRIPTION
**Pull request by ramidecodes**
Adds:
* [x] Update Inference endpoint
* [x] Add new env var for endpoint URL
* [x] Update Chat history mapping for inference payload
* [x] Add System Prompt array
* [x] Add AWS Lambda to trigger inference and expose trigger as API Gateway endpoint

To Do in next PR:
* Add Character name prefix for chat history in prompt
* Fix prompting template structure
* Remove unsafe characters from agent reply
* Listen to "no reply" message from agent

### CODE REVIEW

File: src/server/service/agent.ts [hunk 1]
```diff
@@ -56,6 +69,12 @@ export class Agent {
 
     const message = await this.requestMessageFromLLM();
 
+    // If inference failed or bot decided not to reply, let the agent be silent
+    if (!!!message || message.includes("001001")) {
```
Comment (Mautjee) [inline src/server/service/agent.ts:73]: You commended out the "001001" response in the prompt do you still want to keep it here?

**Review by Mautjee — APPROVED** [review]

File: .env.example [hunk 2]
```diff
@@ -31,3 +31,4 @@ LEADERBOARD_PROGRAM_NAME="leaderboard.aleo"
 
 AWS_ACCESS_KEY="<your_access_key>"
 AWS_SECRET="<your_secret>"
+AWS_INFERENCE_ENDPOINT="https://nez5g29b86.execute-api.eu-central-1.amazonaws.com/staging/generate"
```
[the remark this answers is not in the document]
Comment (silimarius) [reply .env.example:34]: you can also remove all the HUGGING_FACE related variables

File: src/server/service/agent.ts (hunk 1 above)
Comment (ramidecodes) [reply src/server/service/agent.ts:73]: That's coming in the next PR, so no big deal

File: src/types/agent.ts [hunk 3]
```diff
@@ -0,0 +1,12 @@
+import { z } from "zod";
+
+// TODO: Fix role type declaration
+const promptMessageSchema = z.object({
+  role: z.enum(["system", "user", "assistant"]),
+  //   characterName: z.enum(["hal", "dot", "ash", "eve", "roy", "host"]),
```
Comment (silimarius) [inline src/types/agent.ts:6]: Is this useful?

File: src/server/service/agent.ts [hunk 4]
```diff
@@ -56,6 +69,12 @@ export class Agent {
...
+      console.log("Response skipped...");
```
[the remark this answers is not in the document]
Comment (ramidecodes) [reply src/server/service/agent.ts:74]: Good idea.

File: src/server/service/agent.ts [hunk 5]
```diff
@@ -131,4 +154,38 @@ export class Agent {
...
+    const hostMessage = messages.shift();
+    const systemPrompt = `
+    <s>[INST] <<SYS>>
+    ${this._systemPrompt}
+    <</SYS>>
+    ${hostMessage?.content} [/INST]
+    `;
+
+    // TODO: Improve prompt construction
+    const promptString = messages.reduce((acc, currentMessage) => {
```
[the remark this answers is not in the document]
Comment (ramidecodes) [reply src/server/service/agent.ts:180]: Yeah it was also not outputting the right format -_-
This is coming in the next PR.

File: .env.example (hunk 2 above)
[the remark this answers is not in the document]
Comment (ramidecodes) [reply .env.example:34]: Good point..

File: infra/inference-lambda.py [hunk 6]
```diff
@@ -0,0 +1,44 @@
+import os
```
[the remark this answers is not in the document]
Comment (ramidecodes) [reply infra/inference-lambda.py:1]: Coming up in the near future

File: src/server/service/agent.ts [hunk 7]
```diff
@@ -131,4 +154,38 @@ export class Agent {
...
+  private getMessageRole(sender: string): SenderRole {
+    if (sender == this.characterId) return "assistant";
+    else return "user";
+  }
+
+  // TODO: fix case where sender is "host"
+  // private getCharacterName(sender: CharacterId): CharacterName {
+  //   const characterName = CHARACTERS[sender]?.name;
+  //   return !!characterName ? characterName : "roy";
+  // }
```
[the remark this answers is not in the document]
Comment (ramidecodes) [reply src/server/service/agent.ts:167]: Yeap we need to inject them in the chat history for inference

File: src/types/agent.ts (hunk 3 above)
Comment (ramidecodes) [reply src/types/agent.ts:6]: I want to use it in the next PR for injecting the right name in the prompt.

File: src/server/service/agent.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (carlos-kryha) [reply src/server/service/agent.ts:177]: fair

**Review by silimarius — APPROVED** [review]

**Pull request merged**